File: src/hsrc/src/head_msg.cpp
#include <ros/ros.h>
#include <opencv2/opencv.hpp>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/Image.h>
#include <geometry_msgs/Point.h>
#include <geometry_msgs/TransformStamped.h>
#include <sensor_msgs/CameraInfo.h>
#include <darknet_ros_msgs/BoundingBox.h>
#include <darknet_ros_msgs/BoundingBoxes.h>

class Tfpoint_Detector {
public:
    Tfpoint_Detector() {
        bbox = darknet_ros_msgs::BoundingBox();
        pub_threshold = ros::param::param("~pub_threshold", 0.40);
        pub = nh.advertise<geometry_msgs::Point>("point_topic", 10);

        //検出されて初めてサブスクライブする方法
        sub_darknet_bbox = nh.subscribe("/darknet_ros0/bounding_boxes", 1, &Tfpoint_Detector::DarknetBboxCallback, this);
    }

    //target detection
    void DarknetBboxCallback(const darknet_ros_msgs::BoundingBoxes::ConstPtr& darknet_bboxs) {
        darknet_ros_msgs::BoundingBox bbox;
        const std::vector<darknet_ros_msgs::BoundingBox>& head_bboxs = darknet_bboxs->bounding_boxes;
        if (head_bboxs.size() != 0) {
            for (int i = 0; i < head_bboxs.size(); i++) {
                if (head_bboxs[i].Class == "bottle" && head_bboxs[i].probability >= pub_threshold) {
                    bbox = head_bboxs[i];
                    class_name = bbox.Class;
                }else{
                    //
                }
            }
        }else{
            //
        }
        
        cam_x = bbox.xmin + (bbox.xmax - bbox.xmin) / 2;
        cam_y = bbox.ymin + (bbox.ymax - bbox.ymin) / 2;
        sub_swich = true;
        sub_cam_depth = nh.subscribe("/hsrb/head_rgbd_sensor/depth_registered/image_raw", 1, &Tfpoint_Detector::DepthCallback, this);
        // ROS_INFO("x = %.2d, y = %.2d", cam_x, cam_y);
    }

    //depth 
    void DepthCallback(const sensor_msgs::Image::ConstPtr& depth_image_data) {
        try {
            if ( sub_swich == true) {
                cv_bridge::CvImagePtr cv_ptr;
                cv_ptr = cv_bridge::toCvCopy(depth_image_data, sensor_msgs::image_encodings::TYPE_32FC1);
                int depth_x = static_cast<int>(cam_x);
                int depth_y = static_cast<int>(cam_y);
                bbox_depth = cv_ptr->image.at<float>(depth_y, depth_x);
                // ROS_INFO("%d, %d, %f", depth_x, depth_y, bbox_depth);
                sub_camera_info = nh.subscribe("/hsrb/head_rgbd_sensor/depth_registered/camera_info", 1, &Tfpoint_Detector::CameraInfoCallback, this);
                sub_swich = false;
            }else{
                sub_cam_depth.shutdown();
                sub_camera_info.shutdown();
            }
        }
        catch (cv_bridge::Exception& e) {
            ROS_ERROR("Cvbridge error: %s", e.what());
        }
    }

    //座標軸変換
    void CameraInfoCallback(const sensor_msgs::CameraInfo::ConstPtr& info_msg) {

        //検出しないときは通さない
        if(cam_x != 0.0 && cam_y != 0.0 || cam_x > cam_y ){
            try {
                crrection_x = cam_x - info_msg->K[2];//320
                crrection_y = cam_y - info_msg->K[5];//280
                z = bbox_depth;
                x1 = z * (crrection_x / info_msg->K[0]);
                y1 = z * (crrection_y / info_msg->K[4]);
                x = x1 * 0.001 ;
                y = y1 * 0.001 + 0.2 * y1 *0.001;
                z = z * 0.001;
                // ROS_INFO("x = %.2f, y = %.2f, z = %.2f", x, y, z);
            }
            catch (...) {
                ROS_WARN("transform_ERROR");
            }

            //topic Publish
            try {
                geometry_msgs::Point point_msg;
                if (0.40 < z && z <= 1.5){          //距離の制限
                    if (-0.4 < x && x < 0.4){       //　幅の制限
                        point_msg.x = x;
                        point_msg.y = y;
                        point_msg.z = z;
                        pub.publish(point_msg);
                        ROS_INFO("class:%s x = %.2f y = %.2f z = %.2f", class_name.c_str(), x, y, z);
                    }else{
                        ROS_WARN("Out of range");
                    }
                }else{
                    ROS_WARN("Distance over");
                }
            }
            catch (...) {
                ROS_WARN("TopicPublish_ERROR");
            }
        }else{
            ROS_WARN("Not Detection");
        }    
    }

private:
    ros::Publisher pub; 
    ros::Subscriber sub_cam_depth;
    ros::Subscriber sub_darknet_bbox;
    ros::Subscriber sub_camera_info;
    ros::NodeHandle nh;
    cv_bridge::CvImagePtr cv_bridge;
    darknet_ros_msgs::BoundingBox bbox;
    double pub_threshold;
    int cam_x, cam_y;
    float bbox_depth, z;
    float crrection_x , crrection_y, x, y;
    float x1, y1;
    std:: string class_name;
    bool sub_swich;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "Tfpoint_detector_node");
    Tfpoint_Detector tfpoint_detector;
    ros::spin();
    return 0;
}

//1116

File: src/hsrc/scripts/head_msg.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import rospy
from sensor_msgs.msg import Image
from geometry_msgs.msg import Point
from darknet_ros_msgs.msg import BoundingBoxes, BoundingBox
from cv_bridge import CvBridge, CvBridgeError
from sensor_msgs.msg import CameraInfo
import cv2

class Detector():
    def __init__(self):
        rospy.init_node("Point_head_topic")
        self.cv_bridge = CvBridge()
        self.bbox = BoundingBox()
        self.m_pub_threshold = rospy.get_param("~pub_threshold", 0.20)
        self.pub = rospy.Publisher("point_head_topic", Point, queue_size=10)

        self.cam_x = 0.0
        self.cam_y = 0.0

        sub_cam_depth     = rospy.Subscriber("/hsrb/head_rgbd_sensor/depth_registered/image_raw", Image, self.DepthCallback)
        # sub_cam_info      = rospy.Subscriber("/hsrb/head_rgbd_sensor/depth_registered/camera_info", CameraInfo, self.CameraInfo_callback)
        sub_darknet_bbox  = rospy.Subscriber("/darknet_ros0/bounding_boxes", BoundingBoxes, self.DarknetBboxCallback)

    #検出
    def DarknetBboxCallback(self, darknet_bboxs):
        bboxs = darknet_bboxs.bounding_boxes
        bbox = BoundingBox()
        if len(bboxs) != 0 :
            for i, bb in enumerate(bboxs) :
                if bboxs[i].Class == 'bottle' and bboxs[i].probability >= self.m_pub_threshold:
                    bbox = bboxs[i]   
                    self.bbox = bbox
                    #中心座標
                    self.cam_x = self.bbox.xmin + (self.bbox.xmax - self.bbox.xmin) / 2
                    self.cam_y = self.bbox.ymin + (self.bbox.ymax - self.bbox.ymin) / 2
                    # rospy.loginfo("cam= %.2f %.2f ", self.cam_x,self.cam_y)
                elif bboxs[i].Class == 'cup' and bboxs[i].probability >= self.m_pub_threshold:
                    bbox = bboxs[i]
                    self.bbox = bbox
                    #中心座標
                    self.cam_x = self.bbox.xmin + (self.bbox.xmax - self.bbox.xmin) / 2
                    self.cam_y = self.bbox.ymin + (self.bbox.ymax - self.bbox.ymin) / 2
                    # rospy.loginfo("cam= %.2f %.2f ", self.cam_x,self.cam_y)
                else :
                    pass       
                    
    def DepthCallback(self, depth_image_data):
        try:
            #depth距離測定
            self.depth_image = self.cv_bridge.imgmsg_to_cv2(depth_image_data, "passthrough")
            depth_x = int(self.cam_x)
            depth_y = int(self.cam_y)
            self.bbox_depth = self.depth_image[depth_y][depth_x]

            self.CameraInfo_callback()
            # cv2.imshow('depth',self.depth_image)
            # cv2.waitKey(1)

        except CvBridgeError as e:
            rospy.logerr("Cvbridge error: %s", e)

    #座標変換
    # def CameraInfo_callback(self, info_msg):
    def CameraInfo_callback(self):
        if self.cam_x != 0.0 and self.cam_y != 0.0 or self.cam_x > self.cam_y :
            if self.cam_y > 125.0:  #y座標の上の方は対象外
                try:
                    self.x = self.cam_x - 320.7234912485017  #depth_infoの値を入力
                    self.y = self.cam_y - 242.2827148742709
                    self.z = self.bbox_depth

                    self.x1 = self.z * self.x / 533.8084805746594
                    self.y1 = self.z * self.y / 534.057713759154

                    self.x1 = self.x1 * 0.001
                    if self.bbox.Class == "bottle" :
                        self.y1 = self.y1 * 0.001 + 0.2 * self.y1 *0.001
                    else :
                        self.y1 = self.y1 * 0.001
                        
                    self.z = self.z * 0.001
                except:
                    rospy.logwarn("transform_ERROR")

                #topic_publish
                try:
                    if 0.40 < self.z < 1.4 :
                        if -0.4 < self.x1 < 0.4 :
                            if self.bbox.Class == "bottle" :
                                point_msg = Point()
                                point_msg.x = self.x1
                                point_msg.y = self.y1
                                point_msg.z = self.z
                                self.pub.publish(point_msg)
                                rospy.loginfo("x =%.2f y=%.2f z=%.2f ", self.x1, self.y1, self.z)
                            else :
                                point_msg = Point()
                                point_msg.x = self.x1
                                point_msg.y = self.y1
                                point_msg.z = self.z 
                                self.pub.publish(point_msg)
                                rospy.loginfo("x =%.2f y=%.2f z=%.2f ", self.x1, self.y1, self.z)
                        else:
                            rospy.logwarn("Out of range_X")
                    else:
                        rospy.logwarn("Not Detection --> Distance over z=%.2f ", self.z)
                        # rospy.loginfo("x =%.2f y=%.2f z=%.2f ", self.x1, self.y1, self.z)
                        self.cam_x = 0.0
                        self.cam_y = 0.0
                        self.x1 = 0.0
                        self.y1 = 0.0
                        # rospy.logwarn("Distance over")    
                except:
                    rospy.logwarn("Publish_ERROR")
            else:
                rospy.logwarn("Out of range_Y")
        else:
            rospy.logwarn("Not Detection --> ")
            # rospy.loginfo("cam= %.2f %.2f x,y= %.2f %.2f", self.cam_x,self.cam_y, self.x1,self.y1)
            pass

if __name__=="__main__":
    rospy.loginfo("Unable to create tf")
    try:
        
        Detector()
        rospy.spin()

    except rospy.ROSInterruptException:
        pass

#1117
#1120